configure.ac

```diff
@@ -219,24 +219,39 @@ AC_CHECK_LIB([gsl],[gsl_sf_bessel_Jn], [],
 # Check for HDF5 library
 AC_ARG_WITH(hdf5, [AC_HELP_STRING([--without-hdf5],[do not allow HDF5 output])], ok=$withval, ok=yes)
 if test "$ok" = "yes"; then
-	AC_CHECK_LIB(z,deflate, [],[AC_MSG_WARN([zlib is required for HDF5!])])
-
-	save_CC=$CC
-	if test "x$with_mpi" = "xyes"; then
-	   CC=$MPICXX
-	fi
-	save_LIBS_0="$LIBS" # save, to check later if we found any library
-	AC_CHECK_LIB(hdf5, H5Pcreate, [
-		AC_CHECK_HEADERS(hdf5.h, [LIBS="-lhdf5 $LIBS"
-                     AC_DEFINE(HAVE_HDF5,1,[Define if we have & link HDF5])])])
-	if test x"$save_LIBS_0" = x"$LIBS"; then
-		AC_MSG_WARN([Couldn't find the HDF5 library!!  Switching to --without-hdf5.])
-	fi
-	CC=$save_CC
-
-        if test "x$with_mpi" = "xyes"; then
-		AC_CHECK_FUNCS(H5Pset_mpi H5Pset_fapl_mpio)
-	fi
+    AC_CHECK_LIB(z,deflate, [],[AC_MSG_WARN([zlib is required for HDF5!])])
+
+    save_CC="$CC"
+    save_CPPFLAGS="$CPPFLAGS"
+    save_LIBS="$LIBS"
+
+    # Attempt to locate HDF5 via pkg-config and fall back to a default.
+    # Optimistically update LIBS and CPPFLAGS. Revert if any AC_CHECKs fail.
+    PKG_CHECK_MODULES(HDF5, hdf5, [
+        CPPFLAGS="$CPPFLAGS $HDF5_CFLAGS"
+        LIBS="$HDF5_LIBS $LIBS"
+    ], [
+        LIBS="-lhdf5 $LIBS"
+    ])
+
+    if test "x$with_mpi" = "xyes"; then
+        CC=$MPICXX
+        AC_CHECK_FUNCS(H5Pset_mpi H5Pset_fapl_mpio)
+    fi
+
+    AC_CHECK_LIB(hdf5, H5Pcreate, [
```

Comment:
`AC_CHECK_LIB` adds `-lhdf5`, which you don't want to do if you've already added the library above.  You want `AC_CHECK_FUNC` instead.

Reply:
The default action of adding `-lhdf5` is only taken if *action-if-found* is not specified. Here, we are providing an *action-if-found*.

https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Libraries.html